services/docker/docker-size.service.js

```diff
@@ -73,30 +114,76 @@ export default class DockerSize extends BaseJsonService {
     })
   }
 
-  transform({ tag, sort, data }) {
-    if (!tag && sort === 'date') {
-      if (data.count === 0) {
-        throw new NotFound({ prettyMessage: 'repository not found' })
+  noTagWithDateSortTransform(data, arch) {
+    if (data.count === 0) {
+      throw new NotFound({ prettyMessage: 'repository not found' })
+    } else {
+      const latestEntry = data.results[0]
+
+      if (arch) {
+        return { size: getImageForArch(latestEntry.images, arch) }
       } else {
-        return { size: data.results[0].full_size }
+        return { size: latestEntry.full_size }
       }
-    } else if (!tag && sort === 'semver') {
-      const [matches, versions] = data.reduce(
-        ([m, v], d) => {
-          m[d.name] = d.full_size
-          v.push(d.name)
-          return [m, v]
-        },
-        [{}, []]
-      )
-      const version = latest(versions)
+    }
+  }
+
+  noTagWithSemverSortTransform(data, arch) {
+    // If no tag is specified, and sorting is by semver, first filter out the entry containing the latest semver from the response with Docker images.
+    // Then check if any of the returned images for this entry has an architecture matching the arch parameter supplied by the user.
+    // If yes, return the size of the image with this arch.
+    // If not, return the value of the `full_size` from the entry matching the latest semver.
+
+    const [matches, versions, images] = data.reduce(
+      ([m, v, i], d) => {
+        m[d.name] = d.full_size
+        v.push(d.name)
+        i[d.name] = d.images
+        return [m, v, i]
+      },
+      [{}, [], {}]
+    )
+
+    const version = latest(versions)
+
+    let sizeOfImgWithArch
+
+    if (arch) {
+      Object.keys(images).forEach(ver => {
+        if (ver === version) {
+          sizeOfImgWithArch = getImageForArch(images[ver], arch)
+          return { size: sizeOfImgWithArch }
+        }
+      })
+      return { size: sizeOfImgWithArch }
+    } else {
       return { size: matches[version] }
+    }
+  }
+
+  yesTagTransform(data, arch) {
+    // If the tag is specified, check if any of the returned images has an architecture matching the arch parameter supplied by the user.
+    // If yes, return the size of the image with this arch.
+    // If not, return the value of the `full_size` from the response (the image with the `latest` tag).
+    // console.log('yesTagTransform data', data)
+    if (arch) {
+      return { size: getImageForArch(data.images, arch) }
     } else {
       return { size: data.full_size }
     }
   }
 
-  async handle({ user, repo, tag }, { sort }) {
+  transform({ tag, sort, data, arch }) {
+    if (!tag && sort === 'date') {
+      return this.noTagWithDateSortTransform(data, arch)
+    } else if (!tag && sort === 'semver') {
+      return this.noTagWithSemverSortTransform(data, arch)
+    } else {
+      return this.yesTagTransform(data, arch)
+    }
```

Comment:
These 3 functions are all a bit strangely named. In general, the name of a function should describe _what the function does_ not _when we call it_. So if you think about the fragment:

```js
if (!tag && sort === 'date') {
  return this.noTagWithDateSortTransform(data, arch)
```

this is essentially writing the same information that is in the `if` condition (which we can already read - `not tag`, `sort === 'date'`) again as the function name. It doesn't really describe what action is being performed by the function. Naming things is hard, but I'd suggest something like

`noTagWithDateSortTransform()` --> `getSizeFromImageByLatestDate()`
`noTagWithSemverSortTransform()` --> `getSizeFromImageByLatestSemver()`
`yesTagTransform()` --> `getSizeFromTag()`

might be an improvement.

Reply:
Thanks, done.